app/views/courses/TeacherClassesView.coffee

```diff
@@ -264,24 +310,31 @@ module.exports = class TeacherClassesView extends RootView
     window.tracker?.trackEvent $(e.target).data('event-action'), category: 'Teachers', classroomID: classroomID, ['Mixpanel']
     application.router.navigate("/teachers/classes/#{classroomID}", { trigger: true })
 
-  addFreeCourseInstances: ->
+  addFreeCourseInstances: co.wrap ->
     # so that when students join the classroom, they can automatically get free courses
     # non-free courses are generated when the teacher first adds a student to them
-    for classroom in @classrooms.models
-      for course in @courses.models
-        continue if not course.get('free')
-        courseInstance = @courseInstances.findWhere({classroomID: classroom.id, courseID: course.id})
-        if not courseInstance
-          courseInstance = new CourseInstance({
-            classroomID: classroom.id
-            courseID: course.id
-          })
-          # TODO: figure out a better way to get around triggering validation errors for properties
-          # that the server will end up filling in, like an empty members array, ownerID
-          courseInstance.save(null, {validate: false})
-          @courseInstances.add(courseInstance)
-          @listenToOnce courseInstance, 'sync', @addFreeCourseInstances
-          return
+    try
+      promises = []
+      for classroom in @classrooms.models
+        for course in @courses.models
+          continue if not course.get('free')
+          courseInstance = @courseInstances.findWhere({classroomID: classroom.id, courseID: course.id})
+          if not courseInstance
+            courseInstance = new CourseInstance({
+              classroomID: classroom.id
+              courseID: course.id
+            })
+            # TODO: figure out a better way to get around triggering validation errors for properties
+            # that the server will end up filling in, like an empty members array, ownerID
+            promises.push(yield new Promise(courseInstance.save(null, {validate: false}).then))
```

Comment:
Do you need the yield statement in this line? Won't it yield at the `Promise.all` below?

Reply:
yeah thats correct, will update.